beefy-primitives/src/commitment.rs

```diff
@@ -95,6 +99,127 @@ impl<TBlockNumber, TPayload, TSignature> SignedCommitment<TBlockNumber, TPayload
 	}
 }
 
+/// Type to be used to denote placement of signatures
+type BitField = Vec<u8>;
+/// Compress 8 bit values into a single u8 Byte
+const CONTAINER_BIT_SIZE: usize = 8;
+
+/// Compressed representation of [`SignedCommitment`], used for encoding efficiency.
+#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode)]
+struct CompactSignedCommitment<TCommitment, TSignature> {
+	/// The commitment signatures, as is.
+	commitment: TCommitment,
+	/// A `Vec<u8>`, where each bit in a `u8` represents an `Option<TSignature>`
+	/// in the uncompressed GRANDPA validators' signatures.
+	signatures_from: BitField,
+	/// Length of the `Vec` containing uncompressed GRANDPA validators' signatures.
+	signatures_len: u32,
+	/// A `Vec` containing all `TSignature` values present in the uncompressed
+	/// GRANDPA validators' signatures.
+	signatures_compact: Vec<TSignature>,
+}
+
+impl<'a, TBlockNumber, TPayload, TSignature>
+	CompactSignedCommitment<&'a Commitment<TBlockNumber, TPayload>, &'a TSignature>
+{
+	/// Packs a `SignedCommitment` into the compressed `CompactSignedCommitment` format for efficient network transport.
+	fn pack(signed_commitment: &'a SignedCommitment<TBlockNumber, TPayload, TSignature>) -> Self {
+		let SignedCommitment { commitment, signatures } = signed_commitment;
+		let signatures_len = signatures.len() as u32;
+		let mut signatures_from: BitField = vec![];
+		let mut signatures_compact: Vec<&TSignature> = vec![];
+
+		for signature in signatures {
+			match signature {
+				Some(value) => signatures_compact.push(value),
+				None => (),
+			}
+		}
+
+		let bits: Vec<u8> = signatures.iter().map(|x| if x.is_some() { 1 } else { 0 }).collect();
+		let chunks = bits.chunks(CONTAINER_BIT_SIZE);
+		for chunk in chunks {
+			let mut iter = chunk.iter().copied();
+			let mut v = iter.next().unwrap() as u8;
+
+			for bit in iter {
+				v <<= 1;
+				v |= bit as u8;
+			}
+
+			signatures_from.push(v);
+		}
+
+		Self {
+			commitment,
+			signatures_from,
+			signatures_len,
+			signatures_compact,
+		}
+	}
+
+	/// Unpacks a `CompactSignedCommitment` into the uncompressed `SignedCommitment` form.
+	fn unpack(
+		temporary_signatures: CompactSignedCommitment<Commitment<TBlockNumber, TPayload>, TSignature>,
+	) -> SignedCommitment<TBlockNumber, TPayload, TSignature> {
+		let CompactSignedCommitment {
+			commitment,
+			signatures_from,
+			signatures_len,
+			signatures_compact,
+		} = temporary_signatures;
+		let mut bits: Vec<u8> = vec![];
+
+		for block in signatures_from {
+			for bit in 0..CONTAINER_BIT_SIZE {
+				bits.push((block >> (CONTAINER_BIT_SIZE - bit - 1)) & 1);
+			}
+		}
+		let last_byte_len = signatures_len as usize % CONTAINER_BIT_SIZE;
+		let bits = if last_byte_len != 0 {
+			[
+				&bits[..bits.len() - CONTAINER_BIT_SIZE],
+				&bits[bits.len() - last_byte_len..],
+			]
+			.concat()
+		} else {
```

Comment:
We should be avoiding any special casing the last byte, there is definitely a solution that does not involve that.

Basically currently we kind of mess up the bit ordering (numbers are signature indexes):
```
[
  3 2 1 0, // byte 1
  7 6 5 4, /// byte 2
  ...
]
```
This is obviously not wrong, but makes it a bit more difficult to reason about and to propose a proper solution for the last byte issue.

I'd suggest to change the way bits are encoded, so that visually it matches what is in signatures vector. Then after converting to `bits` vector we should make sure the order is correct as well, so that simply truncating works.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9ee8f2399283a346595a205697cd1b04

Did you check this?

Reply:
Oh, missed this somehow but it looks really good, I'd go in this direction.